chain/chain/src/store.rs

```diff
@@ -917,56 +917,187 @@ impl ChainStoreAccess for ChainStore {
+                utils::KeyForAccessKey::new(account_id, access_key_pk).into()
             }
             StateChangesRequest::AllAccessKeyChanges { account_id } => {
-                utils::key_for_all_access_keys(account_id)
+                utils::KeyForAccessKey::get_prefix(account_id).into()
             }
-            StateChangesRequest::CodeChanges { account_id } => utils::key_for_code(account_id),
-            StateChangesRequest::SinglePostponedReceiptChanges { account_id, data_id } => {
-                utils::key_for_postponed_receipt_id(account_id, data_id)
+            StateChangesRequest::CodeChanges { account_id } => {
+                utils::KeyForCode::new(account_id).into()
             }
-            StateChangesRequest::AllPostponedReceiptChanges { account_id } => {
-                utils::key_for_all_postponed_receipts(account_id)
+        };
+        storage_key.extend(&data_key);
+
+        let mut changes_iter =
+            self.store.iter_prefix_ser::<RawStateChangesList>(ColKeyValueChanges, &storage_key);
+        let mut exact_key_changes_iter;
+        let changes_iter: &mut dyn Iterator<Item = _> = match state_changes_request {
+            // These request types should have the key matching exactly
+            StateChangesRequest::AccountChanges { .. }
+            | StateChangesRequest::SingleAccessKeyChanges { .. }
+            | StateChangesRequest::CodeChanges { .. } => {
+                exact_key_changes_iter = changes_iter.filter_map(|change| {
+                    let (key, state_changes) = match change {
+                        Ok(change) => change,
+                        error => {
+                            return Some(error);
+                        }
+                    };
+                    if key.len() != storage_key.len() {
+                        None
+                    } else {
+                        debug_assert_eq!(key, storage_key.as_ref() as &[u8]);
+                        Some(Ok((key, state_changes)))
+                    }
+                });
+                &mut exact_key_changes_iter
             }
-        });
-        let common_key_prefix_len = block_hash.as_ref().len()
-            + match state_changes_request {
-                StateChangesRequest::AccountChanges { .. }
-                | StateChangesRequest::SingleAccessKeyChanges { .. }
-                | StateChangesRequest::AllAccessKeyChanges { .. }
-                | StateChangesRequest::CodeChanges { .. }
-                | StateChangesRequest::SinglePostponedReceiptChanges { .. }
-                | StateChangesRequest::AllPostponedReceiptChanges { .. } => storage_key.len(),
-                StateChangesRequest::DataChanges { account_id, .. } => {
-                    utils::key_for_data(account_id, b"").len()
+
+            StateChangesRequest::AllAccessKeyChanges { .. }
+            | StateChangesRequest::DataChanges { .. } => &mut changes_iter,
+        };
+
+        match state_changes_request {
```

Comment:
I would extract this code into implementations of `From` trait for `StateChangeWithCause` from iterator of state changes, because this method is quite large.

Reply:
Yeah, I will give another round of a try. My initial attempt was not successful (it failed to reduce the amount of code, and only made it worse, but I will see if StateChangeWithCause can save the day).

Reply:
It might not reduce the amount of code but might split it into more encapsulated blocks, improving readability.

Reply:
Also, when we have primitive type A constructible from primitive type B or its collection, then I think we should implement it as `From` trait and keep it next to the declaration of A or B. This seems to be a standard approach in std.